client/transaction-pool/src/lib.rs

```diff
@@ -659,9 +659,10 @@ impl<PoolApi, Block> MaintainedTransactionPool for BasicPool<PoolApi, Block>
 
 						if let Err(e) = pool.resubmit_at(
 							&id,
-							// These transactions are coming from retracted blocks, we should
-							// simply consider them external.
-							TransactionSource::External,
+							// These transactions are coming from retracted blocks, we consider them
+							// as `InBlock` since the given source block must have been previously
+							// imported and validated.
+							TransactionSource::InBlock,
```

Comment:
In the retracted block there could be transactions that were only imported because they were coming from `TransactionSource::Local`.
By importing them as `InBlock` here, you are essentially overriding this and you might import either unsigned transactions or even maybe inherents that shouldn't really be considered by current block author.

Reply:
Do you suggest that we add some other `TransactionSource` variant then? To me there is a distinction between something that came from network/rpc and something that was already in a block and we previously accepted and that got retracted due to consensus (fork choice).